### PR TITLE
app-editors/neovim: depend on and symlink required treesitter parsers

### PR DESCRIPTION
**Pull request by vimproved**
In 0.10.0, neovim made a small subset of tree-sitter parsers (bash, c, lua, python, query, vim, vimdoc) required and load automatically upon opening their respective filetypes. However, this change was missed and was not properly reflected in the 0.10.0 bump, leading to errors upon opening any file of those filetypes (includig running :help). Those parsers have now been added to tree, so this commit adds the correct dependencies and symlinks the parsers to their correct locations in /usr/share/nvim/runtime/parser.

This commit drops keywords, but I do think it's justified because as it currently stands, opening nvim and running :help will lead to an error, which I think is a pretty significant issue. The revbump can be re-keyworded with the new tree-sitter parsers.

---

Please check all the boxes that apply:

- [x] I can submit this contribution in agreement with the [Copyright Policy](https://www.gentoo.org/glep/glep-0076.html#certificate-of-origin).
- [x] I have certified the above via adding a `Signed-off-by` line to *every* commit in the pull request.
- [x] This contribution has not been created with the assistance of Natural Language Processing artificial intelligence tools, in accordance with the [AI policy](https://wiki.gentoo.org/wiki/Project:Council/AI_policy).
- [x] I have run `pkgcheck scan --commits --net` to check for issues with my commits.

Please note that all boxes must be checked for the pull request to be merged.

### CODE REVIEW

**Comment by vimproved** [comment]
Ping on this.

**Review by listout — APPROVED** [review]

**Comment by listout** [comment]
LGTM

**Comment by powerman** [comment]
I'm new to Neovim, just installed it for the first time using neovim-0.10.0-r1.ebuild from this PR.
Here is what I get on `:help` (nvim is executed without any configuration):

```
Error in decoration provider treesitter/highlighter.win:
Error executing lua: /usr/share/nvim/runtime/lua/vim/treesitter/query.lua:252: Query error at 10:3. 
Impossible pattern:                                                                                 
  "~" @markup.heading.4                                                                             
  ^                                                                                                 
                                                                                                    
stack traceback:                                                                                    
        [C]: in function '_ts_parse_query'                                                          
        /usr/share/nvim/runtime/lua/vim/treesitter/query.lua:252: in function 'fn'                  
        /usr/share/nvim/runtime/lua/vim/func/_memoize.lua:58: in function 'fn'                      
        /usr/share/nvim/runtime/lua/vim/func/_memoize.lua:58: in function 'get'                     
        /usr/share/nvim/runtime/lua/vim/treesitter/highlighter.lua:28: in function 'new'            
        /usr/share/nvim/runtime/lua/vim/treesitter/highlighter.lua:240: in function 'get_query'     
        /usr/share/nvim/runtime/lua/vim/treesitter/highlighter.lua:188: in function 'fn'            
        /usr/share/nvim/runtime/lua/vim/treesitter/languagetree.lua:491: in function 'for_each_tree'
                                                                                                    
        /usr/share/nvim/runtime/lua/vim/treesitter/highlighter.lua:175: in function 'prepare_highlig
ht_states'                                                                                          
        /usr/share/nvim/runtime/lua/vim/treesitter/highlighter.lua:398: in function </usr/share/nvim
```

Probably something is still broken in this ebuild.

```
# emerge -pv $(qlist -Iv nvim neovim tree-sitter | sed 's/^/=/')
[ebuild   R    ] dev-libs/tree-sitter-0.22.5:0/0.22.5::gentoo  2964 KiB
[ebuild   R    ] dev-libs/tree-sitter-c-0.21.0::gentoo  USE="-debug -python -test" PYTHON_TARGETS="python3_12 -python3_10 -python3_11" 0 KiB
[ebuild   R   ~] dev-libs/tree-sitter-lua-0.1.0::gentoo  USE="-debug -python -test" PYTHON_TARGETS="python3_12 -python3_10 -python3_11" 0 KiB
[ebuild   R    ] dev-libs/tree-sitter-python-0.21.0-r1::gentoo  USE="-debug -python -test" PYTHON_TARGETS="python3_12 -python3_10 -python3_11" 0 KiB
[ebuild   R   ~] dev-libs/tree-sitter-query-0.4.0::gentoo  USE="-debug -python -test" PYTHON_TARGETS="python3_12 -python3_10 -python3_11" 0 KiB
[ebuild   R   ~] dev-libs/tree-sitter-vim-0.4.0::gentoo  USE="-debug -python -test" PYTHON_TARGETS="python3_12 -python3_10 -python3_11" 0 KiB
[ebuild   R   ~] dev-libs/tree-sitter-vimdoc-3.0.0::gentoo  USE="-debug -python -test" PYTHON_TARGETS="python3_12 -python3_10 -python3_11" 0 KiB
[ebuild   R    ] dev-python/tree-sitter-0.21.3::gentoo  USE="-debug -test" PYTHON_TARGETS="python3_12 -python3_10 -python3_11" 38 KiB
[ebuild   R    ] dev-python/pynvim-0.5.0::gentoo  USE="-test" PYTHON_TARGETS="python3_12 (-pypy3) -python3_10 -python3_11" 0 KiB
[ebuild   R    ] dev-libs/tree-sitter-bash-0.21.0-r1::gentoo  USE="python -debug -test" PYTHON_TARGETS="python3_12 -python3_10 -python3_11" 545 KiB
[ebuild   R   ~] app-editors/neovim-0.10.0-r1::local  USE="nvimpager -test" LUA_SINGLE_TARGET="luajit -lua5-1" 0 KiB

# emerge --info
Portage 3.0.63 (python 3.12.3-final-0, default/linux/amd64/23.0/split-usr/hardened, gcc-13, glibc-2.39-r6, 6.6.30-gentoo x86_64)
=================================================================
System uname: Linux-6.6.30-gentoo-x86_64-AMD_Ryzen_9_5900X_12-Core_Processor-with-glibc2.39
KiB Mem:    32803176 total,   2564188 free
KiB Swap:   16776896 total,  16477888 free
Timestamp of repository gentoo: Wed, 26 Jun 2024 09:30:00 +0000
Head commit of repository gentoo: 1e91fddc4027880e60dc29faf64fa6b83a875fc7
Timestamp of repository guru: Wed, 26 Jun 2024 06:33:57 +0000
Head commit of repository guru: 4d39cddfbab6a24fdcd3a5f60b66a5fd3212eb01

Timestamp of repository steam-overlay: Mon, 17 Jun 2024 09:48:55 +0000
Head commit of repository steam-overlay: 35b2e4df390d79636ff6480e802f556814085aa2

Timestamp of repository torbrowser: Fri, 21 Jun 2024 17:04:09 +0000
Head commit of repository torbrowser: 2ba199fbe61ff5cf12b808342defe69b71ae6ef9

sh bash 5.2_p26-r5
ld GNU ld (Gentoo 2.42 p3) 2.42.0
ccache version 4.9.1 [enabled]
app-misc/pax-utils:        1.3.7::gentoo
app-shells/bash:           5.2_p26-r5::gentoo
dev-build/autoconf:        2.13-r8::gentoo, 2.71-r7::gentoo
dev-build/automake:        1.16.5-r2::gentoo
dev-build/cmake:           3.28.5::gentoo
dev-build/libtool:         2.4.7-r4::gentoo
dev-build/make:            4.4.1-r1::gentoo
dev-build/meson:           1.4.1::gentoo
dev-java/java-config:      2.3.3-r1::gentoo
dev-lang/perl:             5.38.2-r3::gentoo
dev-lang/python:           3.11.9-r1::gentoo, 3.12.3-r1::gentoo
dev-lang/rust:             1.77.1::gentoo
dev-util/ccache:           4.9.1-r1::gentoo
sys-apps/baselayout:       2.15::gentoo
sys-apps/sandbox:          2.38::gentoo
sys-devel/binutils:        2.42-r1::gentoo
sys-devel/binutils-config: 5.5::gentoo
sys-devel/clang:           17.0.6::gentoo
sys-devel/gcc:             13.2.1_p20240210::gentoo
sys-devel/gcc-config:      2.11::gentoo
sys-devel/lld:             17.0.6::gentoo
sys-devel/llvm:            17.0.6::gentoo
sys-kernel/linux-headers:  6.6-r1::gentoo (virtual/os-headers)
sys-libs/glibc:            2.39-r6::gentoo
Repositories:

gentoo
    location: /usr/portage
    sync-type: rsync
    sync-uri: rsync://rsync.gentoo.org/gentoo-portage
    priority: -1000
    volatile: True
    sync-rsync-verify-jobs: 1
    sync-rsync-extra-opts: 
    sync-rsync-verify-metamanifest: yes
    sync-rsync-verify-max-age: 24

guru
    location: /var/db/repos/guru
    sync-type: git
    sync-uri: https://github.com/gentoo-mirror/guru.git
    masters: gentoo
    volatile: False

local
    location: /usr/local/portage
    masters: gentoo
    priority: 0
    volatile: True

steam-overlay
    location: /var/db/repos/steam-overlay
    sync-type: git
    sync-uri: https://github.com/gentoo-mirror/steam-overlay.git
    masters: gentoo
    volatile: False

torbrowser
    location: /var/db/repos/torbrowser
    sync-type: git
    sync-uri: https://github.com/gentoo-mirror/torbrowser.git
    masters: gentoo
    volatile: False

powerman
    location: /home/powerman/proj/gentoo/powerman-overlay
    masters: gentoo
    priority: 50
    volatile: True

Installed sets: @esteam
ACCEPT_KEYWORDS="amd64"
ACCEPT_LICENSE="*"
CBUILD="x86_64-pc-linux-gnu"
CFLAGS="-march=native -O2 -pipe"
CHOST="x86_64-pc-linux-gnu"
CONFIG_PROTECT="/etc /service /usr/lib64/libreoffice/program/sofficerc /usr/share/config /usr/share/easy-rsa /usr/share/gnupg/qualified.txt /usr/share/i2p/scripts /usr/share/maven-bin-3.9/conf /var/log"
CONFIG_PROTECT_MASK="/etc/ca-certificates.conf /etc/dconf /etc/env.d /etc/fonts/fonts.conf /etc/gconf /etc/gentoo-release /etc/revdep-rebuild /etc/sandbox.d /etc/terminfo"
CXXFLAGS="-march=native -O2 -pipe"
DISTDIR="/usr/portage-distfiles"
EMERGE_DEFAULT_OPTS="--with-bdeps=y --autounmask --autounmask-write --alert=y"
ENV_UNSET="CARGO_HOME DBUS_SESSION_BUS_ADDRESS DISPLAY GDK_PIXBUF_MODULE_FILE GOBIN GOPATH PERL5LIB PERL5OPT PERLPREFIX PERL_CORE PERL_MB_OPT PERL_MM_OPT XAUTHORITY XDG_CACHE_HOME XDG_CONFIG_HOME XDG_DATA_HOME XDG_RUNTIME_DIR XDG_STATE_HOME"
FCFLAGS="-march=native -O2 -pipe"
FEATURES="assume-digests binpkg-docompress binpkg-dostrip binpkg-logs buildpkg-live ccache clean-logs config-protect-if-modified distlocks ebuild-locks fixlafiles ipc-sandbox merge-sync merge-wait multilib-strict network-sandbox news parallel-fetch pid-sandbox pkgdir-index-trusted preserve-libs protect-owned qa-unresolved-soname-deps sandbox sfperms strict strict-keepdir unknown-features-warn unmerge-logs unmerge-orphans userfetch userpriv usersandbox usersync warn-on-large-env xattr"
FFLAGS="-march=native -O2 -pipe"
GENTOO_MIRRORS="http://mirror.leaseweb.com/gentoo/ http://mirrors.aliyun.com/gentoo/ http://ftp.snt.utwente.nl/pub/os/linux/gentoo"
LANG="ru_RU.utf8"
LDFLAGS="-Wl,-O1 -Wl,--as-needed -Wl,-z,pack-relative-relocs"
LEX="flex"
LINGUAS="en ru ru_RU"
MAKEOPTS="-j24"
PKGDIR="/usr/portage-packages"
PORTAGE_CONFIGROOT="/"
PORTAGE_RSYNC_OPTS="--recursive --links --safe-links --perms --times --omit-dir-times --compress --force --whole-file --delete --stats --human-readable --timeout=180 --exclude=/distfiles --exclude=/local --exclude=/packages --exclude=/.git"
PORTAGE_TMPDIR="/var/tmp"
SHELL="/bin/zsh"
USE="X a52 aac acl acpi aes alac alsa amd64 avx avx2 bash-completion bluetooth branding bzip2 cairo caps cdda cddb cdr cet chm cli crypt cups dbus dcron dga djvu dri dts dvb dvd dvdr egl eglfs elogind emboss encode exif f16c fam ffmpeg flac fma3 fontconfig gallium gdbm gif gpg gtk hardened heif iconv icu id3tag idn ipv6 jpeg jpeg2k lcms libnotify libtirpc mac mad matroska mmx mmxext mng mp3 mp4 mpeg mtp multilib musepack ncurses network-cron nls nsplugin ogg opengl openmp opus pam pango pclmul pcre pdf perl pic pie png policykit popcnt ppds qt5 rdrand readline rtc sdl seccomp sha spell split-usr sse sse2 sse3 sse4_1 sse4_2 sse4a ssl ssp ssse3 startup-notification svg tcpd test-rust theora tiff truetype udev udisks unicode upower usb vaapi vdpau vim-syntax vorbis wavpack wxwidgets x264 x265 xattr xcb xml xscreensaver xtpax xv xvid xvmc zlib" ABI_X86="64" ADA_TARGET="gcc_12" APACHE2_MODULES="log_config vhost_alias autoindex alias rewrite dir deflate filter mime negotiation auth_basic authn_file authz_host authz_user authz_groupfile cgi actions headers env setenvif authn_core authz_core unixd socache_shmcb access_compat" CALLIGRA_FEATURES="karbon sheets words" COLLECTD_PLUGINS="df interface irq load memory rrdtool swap syslog" CPU_FLAGS_X86="aes avx avx2 f16c fma3 mmx mmxext pclmul popcnt rdrand sha sse sse2 sse3 sse4_1 sse4_2 sse4a ssse3" ELIBC="glibc" GPSD_PROTOCOLS="ashtech aivdm earthmate evermore fv18 garmin garmintxt gpsclock greis isync itrax mtk3301 ntrip navcom oceanserver oncore rtcm104v2 rtcm104v3 sirf skytraq superstar2 tsip tripmate tnt ublox" GRUB_PLATFORMS="efi-64 pc" INPUT_DEVICES="evdev" KERNEL="linux" L10N="en ru" LCD_DEVICES="bayrad cfontz glk hd44780 lb216 lcdm001 mtxorb text" LUA_SINGLE_TARGET="lua5-1" LUA_TARGETS="lua5-1" NGINX_MODULES_HTTP="access auth_basic autoindex browser charset empty_gif fastcgi geo gzip limit_conn limit_req map memcached proxy referer rewrite scgi split_clients ssi upstream_ip_hash userid uwsgi addition fancyindex stub_status" NGINX_MODULES_STREAM="ssl_preread map" OFFICE_IMPLEMENTATION="libreoffice" PHP_TARGETS="php5-6 php7-0" POSTGRES_TARGETS="postgres15" PYTHON_SINGLE_TARGET="python3_12" PYTHON_TARGETS="python3_12" QEMU_SOFTMMU_TARGETS="x86_64 i386" QEMU_USER_TARGETS="x86_64 i386" RUBY_TARGETS="ruby31 ruby32" VIDEO_CARDS="nvidia nouveau" XTABLES_ADDONS="quota2 psd pknock lscan length2 ipv4options ipp2p iface geoip fuzzy condition tarpit sysrq proto logmark ipmark dhcpmac delude chaos account"
Unset:  ADDR2LINE, AR, ARFLAGS, AS, ASFLAGS, CC, CCLD, CONFIG_SHELL, CPP, CPPFLAGS, CTARGET, CXX, CXXFILT, ELFEDIT, EXTRA_ECONF, F77FLAGS, FC, GCOV, GPROF, INSTALL_MASK, LC_ALL, LD, LFLAGS, LIBTOOL, MAKE, MAKEFLAGS, NM, OBJCOPY, OBJDUMP, PORTAGE_BINHOST, PORTAGE_BUNZIP2_COMMAND, PORTAGE_COMPRESS, PORTAGE_COMPRESS_FLAGS, PORTAGE_RSYNC_EXTRA_OPTS, PYTHONPATH, RANLIB, READELF, RUSTFLAGS, SIZE, STRINGS, STRIP, YACC, YFLAGS
```

**Comment by powerman** [comment]
Also, googling this error lead me to https://github.com/nvim-treesitter/nvim-treesitter/issues/3092 which mention `:TSUpdate` command, but trying it result in "E492: Not an editor command: TSUpdate".

`:checkhealth` see no issues in general, for example `:checkhealth vim.treesitter` output:
```
vim.treesitter: require("vim.treesitter.health").check()

- Nvim runtime ABI version: 14
- OK Parser: bash                 ABI: 14, path: /usr/share/nvim/runtime/parser/bash.so
- OK Parser: c                    ABI: 14, path: /usr/share/nvim/runtime/parser/c.so
- OK Parser: lua                  ABI: 14, path: /usr/share/nvim/runtime/parser/lua.so
- OK Parser: python               ABI: 14, path: /usr/share/nvim/runtime/parser/python.so
- OK Parser: query                ABI: 14, path: /usr/share/nvim/runtime/parser/query.so
- OK Parser: vim                  ABI: 14, path: /usr/share/nvim/runtime/parser/vim.so
- OK Parser: vimdoc               ABI: 14, path: /usr/share/nvim/runtime/parser/vimdoc.so
```

**Comment by vimproved** [comment]
> ```
> Error in decoration provider treesitter/highlighter.win:
> Error executing lua: /usr/share/nvim/runtime/lua/vim/treesitter/query.lua:252: Query error at 10:3. 
> Impossible pattern:                                                                                 
>   "~" @markup.heading.4                                                                             
>   ^                                                                                                 
>                                                                                                     
> stack traceback:                                                                                    
>         [C]: in function '_ts_parse_query'                                                          
>         /usr/share/nvim/runtime/lua/vim/treesitter/query.lua:252: in function 'fn'                  
>         /usr/share/nvim/runtime/lua/vim/func/_memoize.lua:58: in function 'fn'                      
>         /usr/share/nvim/runtime/lua/vim/func/_memoize.lua:58: in function 'get'                     
>         /usr/share/nvim/runtime/lua/vim/treesitter/highlighter.lua:28: in function 'new'            
>         /usr/share/nvim/runtime/lua/vim/treesitter/highlighter.lua:240: in function 'get_query'     
>         /usr/share/nvim/runtime/lua/vim/treesitter/highlighter.lua:188: in function 'fn'            
>         /usr/share/nvim/runtime/lua/vim/treesitter/languagetree.lua:491: in function 'for_each_tree'
>                                                                                                     
>         /usr/share/nvim/runtime/lua/vim/treesitter/highlighter.lua:175: in function 'prepare_highlig
> ht_states'                                                                                          
>         /usr/share/nvim/runtime/lua/vim/treesitter/highlighter.lua:398: in function </usr/share/nvim
> ```

urgh, this seems to be caused by a version mismatch between the version of `dev-libs/tree-sitter-vimdoc` in tree (3.0.0) and the version used by neovim (2.5.1). 2.5.1 isn't in tree so we'll need to add that and pin the dependency.

**Comment by vimproved** [comment]
Not the happiest with this fix. Nevertheless, it works.

**Comment by powerman** [comment]
Works for me, thanks!

**Comment by listout** [comment]
@vimproved good work, thank you.

File: app-editors/neovim/neovim-0.10.0-r1.ebuild [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+	>=dev-libs/msgpack-3.0.0:=
+	>=dev-libs/tree-sitter-0.20.9:=
+	>=dev-libs/tree-sitter-bash-0.21 <dev-libs/tree-sitter-bash-0.22
+	>=dev-libs/tree-sitter-c-0.21 <dev-libs/tree-sitter-c-0.22
+	>=dev-libs/tree-sitter-lua-0.1 <dev-libs/tree-sitter-lua-0.2
+	>=dev-libs/tree-sitter-markdown-0.2 <dev-libs/tree-sitter-markdown-0.3
+	>=dev-libs/tree-sitter-python-0.21 <dev-libs/tree-sitter-python-0.22
+	>=dev-libs/tree-sitter-query-0.3 <dev-libs/tree-sitter-query-0.4
+	>=dev-libs/tree-sitter-vim-0.4 <dev-libs/tree-sitter-vim-0.5
+	>=dev-libs/tree-sitter-vimdoc-2 <dev-libs/tree-sitter-vimdoc-3
```
Comment (eli-schwartz) [inline app-editors/neovim/neovim-0.10.0-r1.ebuild:63]: Maybe use `=cat/pkg-1.2*` style dependencies instead, given this explicitly pins to semver?

**Review by arthurzam — APPROVED** [review]
tested, and it indeed works.

(my experience with neovim is null, so I just emerged it based on current PR, and tries to open various files with `nvim` and they looks like correct code highlighting)

File: app-editors/neovim/neovim-9999.ebuild [hunk 2]
```diff
@@ -52,6 +53,14 @@ DEPEND="${LUA_DEPS}
...
 	>=dev-libs/msgpack-3.0.0:=
 	>=dev-libs/tree-sitter-0.22.6:=
+	>=dev-libs/tree-sitter-bash-0.21 <dev-libs/tree-sitter-bash-0.22
+	>=dev-libs/tree-sitter-c-0.21 <dev-libs/tree-sitter-c-0.22
+	>=dev-libs/tree-sitter-lua-0.1 <dev-libs/tree-sitter-lua-0.2
+	>=dev-libs/tree-sitter-markdown-0.2 <dev-libs/tree-sitter-markdown-0.3
+	>=dev-libs/tree-sitter-python-0.21 <dev-libs/tree-sitter-python-0.22
+	>=dev-libs/tree-sitter-query-0.4 <dev-libs/tree-sitter-query-0.5
+	>=dev-libs/tree-sitter-vim-0.4 <dev-libs/tree-sitter-vim-0.5
+	>=dev-libs/tree-sitter-vimdoc-3 <dev-libs/tree-sitter-vimdoc-4
```
Comment (arthurzam) [inline app-editors/neovim/neovim-9999.ebuild:63]: please also update here to the suggested syntax above by @eli-schwartz

**Comment by powerman** [comment]
Also tested.
Both ebuilds works with `LUA_SINGLE_TARGET="luajit"` and both fails to build with `LUA_SINGLE_TARGET="lua5-1"`.

**Comment by powerman** [comment]
TBH I'm not sure is it makes sense to support lua5-1 - it's much slower so chances are everyone wants Neovim with luajit anyway. But it also affects dev-lua/luv, so in case someone needs dev-lua/luv with lua5-1 he'll also have to use Neovim with lua5-1.

**Comment by mattst88** [comment]
> TBH I'm not sure is it makes sense to support lua5-1 - it's much slower so chances are everyone wants Neovim with luajit anyway. But it also affects dev-lua/luv, so in case someone needs dev-lua/luv with lua5-1 he'll also have to use Neovim with lua5-1.

I guess there's some value in keeping `lua5-1` support for architectures that don't have luajit support.

**Review by juippis — APPROVED** [review]
I hit this https://bugs.gentoo.org/932671 but the editor seems to work.

**Comment by juippis** [comment]
`:help` works in 0.10.0-r1 but not in -9999 anymore.

**Pull request closed without merging**